_posts/2025-12-23-hibernate-panache-next.asciidoc

```diff
@@ -0,0 +1,253 @@
+to be blocked having to choose which entity type to extend, between the blocking `PanacheEntity` and its non-blocking
+counterpart.
+- Both Hibernate ORM with Panache and its Hibernate Reactive counterpart only suppose managed entities, and we want
+to add support for stateless entities (using `StatelessSession`), but we cannot do that with the current Panache API.
+- We want to add support for linkg:https://jakarta.ee/specifications/data/1.0/jakarta-data-1.0[Jakarta Data],
+which allows defining type-safe queries using annotations, which are
+checked at build time, but the current
+link:https://quarkus.io/guides/hibernate-orm-panache#solution-1-using-the-active-record-pattern[_active record_]
+version of Panache entities does not easily support this.
+- We currently have a split between _active record_ and
+link:https://quarkus.io/guides/hibernate-orm-panache#solution-2-using-the-repository-pattern[_repository_] modes of
+operations, which is not satisfactory, and can be confusing to new users.
+- The static methods of _active record_ entities have a hole in their types which require they be either assigned to
+known types (not `var`) or have their type arguments explicitly specified.
+- Method pointers cannot be used on the static methods of _active record_ entities, causing confusion.
+
+I could spend a long time explaining each of these issues, and the many things we experimented on to fix them, but
+you're probably more interested in what we came up with.
+
+# So, what's this new API?
+
+It's a new Quarkus extension, and it lets you specify your entity in a way which allows you to use it using a blocking
+JDBC driver, or a Reactive driver, as a managed entity, or as a stateless entity, with type-safe queries, or non-type-safe
+queries. It's very flexible and it's still clearly a continuation of the old Panache APIs, with a dash of
+linkg:https://jakarta.ee/specifications/data/1.0/jakarta-data-1.0[Jakarta Data]
+and all mixed in a way that we hope you will like.
+
+https://quarkus.io/version/main/guides/hibernate-panache[The documentation] has a lot of
+information as to how to set it up, and all the options, so I will focus on some
+bite-sized samples for this blog post. Here's your first entity with type-safe queries:
+
+```java
+@Entity
+public class Cat extends PanacheEntity {
+    public String name;
+    public LocalDate birth;
+    public Breed breed;
+
+    public enum Breed {
+        CUTE, HAIRLESS;
+    }
+
+    public interface Repo extends PanacheRepository<Cat> {
+        @Find
+        Cat findByName(String name);
+
+        @HQL("where breed = CUTE")
+        List<Cat> findCute();
+
+        @HQL("delete from Cat where name = :name")
+        long deleteByName(String name);
+
+        @HQL("delete from Cat where breed = HAIRLESS")
+        long deleteHairless();
+    }
+}
+```
+
+Here's what you can notice right away:
+
+- Your entity extends `PanacheEntity` (a new type), which means: by default it's blocking and managed.
```

Comment:
I remember this already exists in Panache 1 in the `io.quarkus.hibernate.orm.panache` package. What makes it a new type? Another package?

Reply:
Yes.